[fix] Ensure recent items are shown first in the REST API

stktyagi opened the pull request:
## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [ ] I have updated the documentation.

## Reference to Existing Issue

Closes #1040.

## Description of Changes

Added ordering by creation date (descending) to the Base command view to ensure consistent and expected command ordering and checked all other list apis to ensure recency.
Tests to supported the same added.

stktyagi commented:
@nemesifier 
I've created a new pull request which follows the contributing guidelines.

stktyagi commented:
@nemesifier build was failing due to the commit message so I have updated it kindly review.
thanks!

stktyagi commented:
updated commit message to prevent build failure.

nemesifier reviewed (commented):
@stktyagi it looks good to me! I just want to find some time to double check this is the only endpoint where we need to fix this.

nemesifier commented on openwisp_controller/connection/tests/test_api.py:
```diff
@@ -410,6 +417,18 @@ def test_filter_credentials_list(self):
         self.assertNotContains(response, cred_1.id)
 
+    @patch.object(ListViewPagination, "page_size", 3)
+    def test_credential_list_ordering(self):
```
This could be a subtest of `test_get_credentials_list`, can you please merge this with that test and simplify the logic a bit to avoid bloat? Just create 3 credentials, check if you can set the creation date of one of them manually to an older date and verify that it's shown as last.

nemesifier commented on openwisp_controller/connection/tests/test_api.py:
```diff
@@ -484,6 +503,25 @@ def test_get_deviceconnection_list(self):
         self.assertEqual(response.data['count'], 0)
 
+    @patch.object(ListViewPagination, "page_size", 3)
+    def test_deviceconnection_list_ordering(self):
```
same here, I would run this as a subtest of `test_get_deviceconnection_list` and simplify the logic to avoid bloat.

nemesifier reviewed (commented):
Other endpoints I found that are suffering from the same issue:

- `/api/v1/controller/group/`
- `/api/v1/controller/device/`

Can we take advantage and fix those as well?

Regarding the tests, I recommend extending the existing tests for these endpoints with a subtest which checks the ordering, let's try to reduce the new lines to the bare minimuim required to test the ordering effectively.

Please also update your local copy of openwisp-utils to the latest master, reinstall it to update the dependencies and run `openwisp-qa-format` again as there's been slight changes to the QA dependencies and configuration (see https://github.com/openwisp/openwisp-utils/pull/473 and https://github.com/openwisp/openwisp-utils/pull/456).

See also my comments below.

stktyagi commented:
@nemesifier  greetings,
updated test files with new subtests instead of redundant test functions and updated logic to avoid bloat.
Tested for all endpoints which include 2 of the previous ones, /group and /device.
complied with new qa-tests.

nemesifier commented on openwisp_controller/config/tests/test_api.py:
```diff
@@ -249,6 +252,32 @@ def test_device_list_api(self):
             r = self.client.get(path)
         self.assertEqual(r.status_code, 200)
+        with self.subTest("should return devices ordered by most recent first"):
+            Device.objects.all().delete()
```
Why is this needed? We could avoid it and add only 2 more devices so we end up with 3.

nemesifier reviewed (approved):
I must have made some confusion during the previous review because I see that device and device group are already ordered properly. Well, that's good to know, thanks for adding the tests! I did some minor simplifications, should be ready to merge.
